src/main/java/io/cdap/plugin/cdc/source/salesforce/SalesforceReceiver.java

```diff
@@ -0,0 +1,250 @@
+      try {
+        String message = eventTopicListener.getMessage(GET_MESSAGE_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+
+        if (message != null) {
+          // whole message class is not needed because we are interested only in change event payload
+          JsonObject headerElement = JSON_PARSER.parse(message)
+            .getAsJsonObject()
+            .getAsJsonObject("data")
+            .getAsJsonObject("payload")
+            .getAsJsonObject("ChangeEventHeader");
+          ChangeEventHeader event = GSON.fromJson(headerElement, ChangeEventHeader.class);
+
+          List<ChangeEventHeader> eventsList = events.getOrDefault(event.getTransactionKey(), new ArrayList<>());
+          eventsList.add(event);
+
+          if (event.isTransactionEnd()) {
+            processEvents(eventsList, connection);
+            events.remove(event.getTransactionKey());
+          } else {
+            events.put(event.getTransactionKey(), eventsList);
+          }
+        }
+      } catch (Exception e) {
+        switch (errorHandling) {
+          case SKIP:
+            LOG.warn("Failed to process message, skipping it.", e);
+            break;
+          case STOP:
+            throw new RuntimeException("Failed to process message", e);
+          default:
+            throw new IllegalStateException(String.format("Unknown error handling strategy '%s'", errorHandling));
+        }
+      }
+    }
+    eventTopicListener.stop();
+  }
+
+  private void processEvents(List<ChangeEventHeader> events, PartnerConnection connection) throws ConnectionException {
+    for (ChangeEventHeader event : events) {
+      SObjectDescriptor descriptor = SObjectDescriptor.fromName(event.getEntityName(), connection);
+      SObjectsDescribeResult describeResult = new SObjectsDescribeResult(connection, descriptor.getAllParentObjects());
+
+      Schema schema = SalesforceRecord.getSchema(descriptor, describeResult);
+      updateSchemaIfNecessary(event.getEntityName(), schema);
+
+      if (getOperationType(event) != OperationType.DELETE) {
+        sendUpdateRecords(event, descriptor, schema, connection);
+      } else {
+        sendDeleteRecords(Arrays.asList(event.getRecordIds()), event.getEntityName(), schema);
+      }
+    }
+  }
+
+  private void updateSchemaIfNecessary(String entityName, Schema schema) {
+    Schema previousSchema = schemas.get(entityName);
+
+    if (!schema.equals(previousSchema)) {
+      StructuredRecord ddlRecord = SalesforceRecord.buildDDLStructuredRecord(entityName, schema);
+      schemas.put(entityName, schema);
+
+      LOG.debug("Sending ddl message for '{}'", entityName);
+      store(ddlRecord);
+    }
+  }
+
+  private void sendUpdateRecords(ChangeEventHeader event, SObjectDescriptor descriptor, Schema schema,
+                                 PartnerConnection connection) throws ConnectionException {
+    String query = getQuery(event, descriptor.getFieldsNames());
+    QueryResult queryResult = connection.query(query);
```

Comment:
This is querying the SObject to get the record value?? Maybe I am missing some key information about the salesforce API.

What happens when 2 records with the same key get updated in an SObject? Do we get 2 change events? This type of information needs to be documented.

Reply:
If update is similar for several records of the same type, we can receive the list of ids that were changed and common changes of their fields.